agent/ecscni/plugin.go

```diff
@@ -69,7 +70,29 @@ func NewClient(cfg *Config) CNIClient {
 
 // SetupNS will set up the namespace of container, including create the bridge
 // and the veth pair, move the eni to container namespace, setup the routes
-func (client *cniClient) SetupNS(cfg *Config) (*current.Result, error) {
+func (client *cniClient) SetupNS(ctx context.Context, cfg *Config) (*current.Result, error) {
+	type output struct {
+		result *current.Result
+		err    error
+	}
+	response := make(chan output)
+	go func(response chan output) {
+		result, err := client.setupNS(cfg)
+		response <- output{
+			result: result,
+			err:    err,
+		}
+	}(response)
+
+	select {
+	case <-ctx.Done():
+		return nil, errors.Wrap(ctx.Err(), "cni setup: container namespace setup failed")
```

Comment:
I think you have to check if `ctx.Err()` is non nil. I don't think you can blindly do it this way. On second thought, I don't really think you need that async routine at all. If you pass in timeout as an arg to this method, you can do it in a way that's similar to what I'm doing in https://github.com/aws/amazon-ecs-agent/pull/1363 and deal with it in a much better way.

Reply:
Based on the document [here](https://golang.org/pkg/context/#Context), I don't think check `ctx.Err()!=nil` is required.
```
// If Done is not yet closed, Err returns nil.
// If Done is closed, Err returns a non-nil error explaining why:
// Canceled if the context was canceled
// or DeadlineExceeded if the context's deadline passed.
// After Err returns a non-nil error, successive calls to Err return the same error.
```

>A much better model would be to create the derived context within this method

Yes, a timeout instead should be passed in instead of passing the context, will modify that.

Reply:
> Based on the document [here](https://golang.org/pkg/context/#Context), I don't think check `ctx.Err()!=nil` is required.

Thanks! make sense.